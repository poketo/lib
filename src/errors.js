// @flow

type ErrorCode =
  | 'INVALID_URL'
  | 'UNSUPPORTED_SITE'
  | 'UNSUPPORTED_SITE_REQUEST'
  | 'TIMEOUT'
  | 'HTTP_ERROR';

class PoketoError extends Error {
  code: ErrorCode;
  constructor(errorCode: ErrorCode, message: string) {
    super(message);
    Error.captureStackTrace(this, this.constructor);
    this.name = this.constructor.name;
    this.code = errorCode;
  }
}

class InvalidUrlError extends PoketoError {
  constructor(url: string) {
    super('INVALID_URL', `Unable to parse url '${url}'`);
  }
}

class HTTPError extends PoketoError {
  statusCode: number;
  url: string;

  constructor(statusCode: number, message: string, url: string) {
    super('HTTP_ERROR', message);
    this.statusCode = statusCode;
    this.url = url;
  }
}

class NotFoundError extends HTTPError {
  constructor(url: string) {
    super(404, 'Not Found', url);
  }
}

class TimeoutError extends PoketoError {
  constructor(message: string, url: string) {
    super('TIMEOUT', message);
  }
}

class UnsupportedSiteError extends PoketoError {
  constructor(site: string) {
    super('UNSUPPORTED_SITE', `Site at '${site}' is not supported`);
  }
}

class UnsupportedSiteRequestError extends PoketoError {
  constructor(siteName: string, operationName: string) {
    super(
      'UNSUPPORTED_SITE_REQUEST',
      `${siteName} does not support ${operationName}`,
    );
  }
}

export default {
  PoketoError,
  HTTPError,
  InvalidUrlError,
  NotFoundError,
  TimeoutError,
  UnsupportedSiteError,
  UnsupportedSiteRequestError,
};
