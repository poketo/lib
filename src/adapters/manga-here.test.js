import poketo from '../index';
import site from './manga-here';
import errors from '../errors';

describe('MangaHereAdapter', () => {
  describe('supportsUrl', () => {
    it('returns true for urls like mangahere.cc or .co', () => {
      expect(site.supportsUrl('http://www.mangahere.cc')).toBe(true);
      expect(site.supportsUrl('http://www.mangahere.co')).toBe(true);
      expect(site.supportsUrl('https://mangahere.com')).toBe(true);
    });

    it('returns false for urls that are not mangahere.cc or .co', () => {
      expect(site.supportsUrl('http://www.mangahere.kc')).toBe(false);
      expect(site.supportsUrl('http://ma.ngahere.co')).toBe(false);
    });
  });

  describe('parseUrl', () => {
    it('returns the components of a url', () => {
      expect(
        site.parseUrl('http://www.mangahere.cc/manga/urami_koi_koi_urami_koi/'),
      ).toEqual({
        seriesSlug: 'urami_koi_koi_urami_koi',
        chapterSlug: null,
      });

      expect(
        site.parseUrl(
          'http://www.mangahere.cc/manga/urami_koi_koi_urami_koi/c038/1.html',
        ),
      ).toEqual({ seriesSlug: 'urami_koi_koi_urami_koi', chapterSlug: 'c038' });

      expect(
        site.parseUrl(
          'http://www.mangahere.cc/manga/flying_witch_ishizuka_chihiro/c018.5/',
        ),
      ).toEqual({
        seriesSlug: 'flying_witch_ishizuka_chihiro',
        chapterSlug: 'c018.5',
      });
    });

    it('throws on unparseable paths', () => {
      expect(() => {
        site.parseUrl('http://www.mangahere.cc/banana');
      }).toThrow(errors.InvalidUrlError);

      expect(() => {
        site.parseUrl('http://www.mangahere.cc/marnga/hello/world');
      }).toThrow(errors.InvalidUrlError);
    });
  });
});

xdescribe('MangaHere', () => {
  const server = new AdapterVcrServer(site);

  beforeAll(async () => {
    await server.listenAndMock(57157);
  });

  afterAll(() => {
    server.close();
  });

  describe('getSeries', () => {
    it('returns a metadata object', async () => {
      const { chapters, ...metadata } = await poketo.getSeries(
        'manga-here:urami_koi_koi_urami_koi',
      );

      expect(metadata).toMatchSnapshot();

      const chapterNumbersToTest = ['13'];
      const chaptersToTest = chapters.filter(chapter =>
        chapterNumbersToTest.includes(chapter.chapterNumber),
      );
      expect(chaptersToTest).toMatchSnapshot();
    });
  });

  describe('getChapter', () => {
    it('returns a list of pages', async () => {
      const chapter = await poketo.getChapter(
        'manga-here:urami_koi_koi_urami_koi:c038',
      );

      expect(chapter.pages).toHaveLength(27);
      expect(chapter.pages[0]).toEqual(
        expect.objectContaining({
          id: expect.any(String),
          url: expect.stringContaining('http'),
        }),
      );
    });
  });
});
