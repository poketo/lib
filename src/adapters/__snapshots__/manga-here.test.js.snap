// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MangaHereAdapter getSeries returns a metadata object 1`] = `
Object {
  "slug": "urami_koi_koi_urami_koi",
  "title": "Urami Koi, Koi, Urami Koi.",
  "url": "http://mangahere.cc/manga/urami_koi_koi_urami_koi",
}
`;

exports[`MangaHereAdapter getSeries returns a metadata object 2`] = `
Array [
  Object {
    "chapterNumber": "13",
    "createdAt": 1449993600,
    "slug": "c013",
    "title": "Love is Like an Ox",
    "url": "http://mangahere.cc/manga/urami_koi_koi_urami_koi/c013",
  },
]
`;
